jax/_src/numpy/lax_numpy.py

```diff
@@ -3706,11 +3706,14 @@ def _index_to_gather(x_shape, idx, normalize_indices=True):
       # Normalize the slice to use None when possible
       start, stop, step = i.start, i.stop, i.step
       try:
-        if ((step is None or core.symbolic_equal_dim(step, 1)) and
-            stop is not None and core.symbolic_equal_dim(stop, x_shape[x_axis])):
-          # The following is a useful special case with shape polymorphism
-          stop = None
-      except TypeError:
+        if step is None or core.symbolic_equal_dim(step, 1):
+          step = None
+        if step is None:
+          if start is None or core.symbolic_equal_dim(start, 0):
+            start = None
+          if stop is None or core.greater_equal_dim(stop, x_shape[x_axis]):
```

Comment:
I am not 100% sure that it is OK to use `greater_equal_dim` here because it includes this special case:
https://github.com/google/jax/blob/35b32eef96e579e5f565c48f5962463f8d82192d/jax/core.py#L1515-L1521

If this could cause problems I am happy switch back to `symbolic_equal_dim` which would mean that indices larger than the size of the axis wouldn't use this fast path.

Reply:
I added an additional check to make sure that `stop` isn't a `Tracer` which should make this work correctly even for a possible dynamic case.